### PR TITLE
Faster sync-fields

### PR DESCRIPTION
**Pull request by calherries**
This is part 2 towards resolving https://github.com/metabase/metabase/issues/38492
Part 1 is https://github.com/metabase/metabase/pull/38970

This PR adds an alternative implementation of the sync-fields step of sync where we sync the foreign key information of a database.

It adds a new driver feature: `describe-fields`. If a driver supports this feature, they need to implement the new driver method `metabase.driver/describe-fields`. If the driver doesn't support the feature, we'll continue to use `metabase.driver/describe-table`. The plan for driver authors to implement `describe-fields` in terms of `driver/describe-table`, which should be a really simple change for most drivers, assuming it doesn't need to be performant for large DBs.

The way this works is instead of querying the customer DB one table at a time using the `getColumns` JDBC method, we execute a single query that gets all the data at once with paginated results. We then reduce over those results, updating our App DB one foreign key at a time. This approach mirrors https://github.com/metabase/metabase/pull/38970

I've tested this with the "massive" redshift instance, and this is how long each sync step took:
- "sync-tables": 13.3 s
- "sync-fields": 1.0 mins (this is the step we're speeding up)
- "sync-fks": 1.3 s (there are no FKs to sync though)
- total time for all sync metadata steps: 5.0 minutes

Before this PR, `sync-fields` took hours (too long to bother measuring precisely anyway).

### CODE REVIEW

File: src/metabase/driver/sql_jdbc/sync/describe_table.clj [hunk 1]
```diff
@@ -121,12 +121,25 @@
          (merge
            {:name                       column-name
             :database-type              (.getString rs "TYPE_NAME")
+            :table-schema               (.getString rs "TABLE_SCHEM")
```
Comment (luizarakaki) [inline src/metabase/driver/sql_jdbc/sync/describe_table.clj:124]: typo?

Comment (calherries) [reply src/metabase/driver/sql_jdbc/sync/describe_table.clj:124]: Nope, these are from the jdbc spec

**Comment by calherries** [comment]
Note to self: remember to clean up this code https://github.com/metabase/metabase/pull/24910/files

File: docs/developers-guide/driver-changelog.md [hunk 2]
```diff
@@ -4,6 +4,14 @@ title: Driver interface changelog
...
 # Driver Interface Changelog
 
+## Metabase 0.50.0
+
+- The multimethod `metabase.driver/describe-fields` has been added. The method needs to be
+  implemented if the database supports the `:fast-sync-fields` feature.
+
+- The multimethod `metabase.driver/describe-fks` has been added. The method needs to be implemented if the database
+  supports the `:foreign-keys` and `:fast-sync-fks` features.
+
```
Comment (crisptrutski) [inline docs/developers-guide/driver-changelog.md:14]: Do you mean for this PR to be stacked on #38970? It seems to include it.

Comment (calherries) [reply docs/developers-guide/driver-changelog.md:14]: yes

**Comment by calherries** [comment]
Note to self: check database-position is not different

**Comment by calherries** [comment]
depends on https://github.com/metabase/metabase/pull/40006

File: docs/developers-guide/driver-changelog.md [hunk 3]
```diff
@@ -6,6 +6,12 @@ title: Driver interface changelog
 
 ## Metabase 0.50.0
```
Comment (qnkhuat) [inline docs/developers-guide/driver-changelog.md:7]: should be 49

File: docs/developers-guide/driver-changelog.md [hunk 4]
```diff
@@ -6,6 +6,12 @@ title: Driver interface changelog
...
 
+- The multimethod `metabase.driver/describe-fields` has been added. The method needs to be
+  implemented if the database supports the `:fast-sync-fields` feature.
+
+- The multimethod `metabase.driver.sql-jdbc.sync.describe-table/describe-fields-sql` has been added. The method needs
+  to be implemented if you want to use the default JDBC implementation of `metabase.driver/describe-fields`.
```
Comment (qnkhuat) [inline docs/developers-guide/driver-changelog.md:13]: should be 49

File: modules/drivers/redshift/src/metabase/driver/redshift.clj [hunk 5]
```diff
@@ -359,18 +399,6 @@
...
-  [driver conn {schema :schema, table-name :name :as table} db-name-or-nil]
-  (let [parent-method (get-method sql-jdbc.describe-table/describe-table-fields :sql-jdbc)]
-    (try (parent-method driver conn table db-name-or-nil)
-         (catch Exception e
-           (log/error e (trs "Error fetching field metadata for table {0}" table-name))
-           ;; Use the fallback method (a SELECT * query) if the JDBC driver throws an exception (#21215)
-           (into
-            #{}
-            (sql-jdbc.describe-table/describe-table-fields-xf driver table)
-            (sql-jdbc.describe-table/fallback-fields-metadata-from-select-query driver conn db-name-or-nil schema table-name))))))
```
Comment (calherries) [inline modules/drivers/redshift/src/metabase/driver/redshift.clj:372]: This fallback was added as part of https://github.com/metabase/metabase/pull/24910. 

The test for that issue, `redshift-lbv-sync-error-test`, passes without the fallback now, so I believe the root cause of the issue was with how we were calling `getColumns`.

File: test/metabase/driver/sql_jdbc/sync/describe_table_test.clj [hunk 6]
```diff
@@ -126,8 +149,31 @@
            {:fk-column-name "VENUE_ID", :dest-table {:name "VENUES", :schema "PUBLIC"}, :dest-column-name "ID"}}
          (sql-jdbc.describe-table/describe-table-fks :h2 (mt/id) {:name "CHECKINS"}))))
 
+(deftest describe-fields-or-table-test
```
Comment (calherries) [inline test/metabase/driver/sql_jdbc/sync/describe_table_test.clj:152]: I didn't find any test coverage for `:database-position` for the default SQL-JDBC implementation so I added this test. It also sense checks a few other things, like sensible base types and names.

File: src/metabase/driver/postgres.clj
Comment (qnkhuat) [inline src/metabase/driver/postgres.clj:1]: why do we have these changes?

File: test/metabase/driver/sql_jdbc_test.clj [hunk 7]
```diff
@@ -79,6 +80,19 @@
          #_{:clj-kondo/ignore [:deprecated-var]}
          (driver/describe-table-fks :h2 (mt/db) (t2/select-one Table :id (mt/id :venues))))))
 
+(deftest describe-fields-sync-with-composite-pks-test
```
Comment (calherries) [inline test/metabase/driver/sql_jdbc_test.clj:83]: I couldn't find any test coverage for composite PKs so I added this test.

File: src/metabase/sync/interface.clj [hunk 8]
```diff
@@ -65,16 +66,15 @@
...
-   [:name                         ::lib.schema.common/non-blank-string]
-   [:schema                       [:maybe ::lib.schema.common/non-blank-string]]
-   [:fields                       [:set TableMetadataField]]
-   [:description {:optional true} [:maybe :string]]])
-
-(def TableMetadata
-  "Schema for the expected output of [[metabase.driver/describe-table]]."
-  [:ref ::TableMetadata])
+(mr/def ::FieldMetadataEntry
+  (-> (mr/schema ::TableMetadataField)
```
Comment (qnkhuat) [inline src/metabase/sync/interface.clj:70]: why did you remove description? and any reason why you use these helpers instead of the normal syntax?

File: modules/drivers/redshift/test/metabase/driver/redshift_test.clj [hunk 9]
```diff
@@ -241,8 +241,8 @@
                view-nm))
           (let [table-id (t2/select-one-pk Table :db_id (u/the-id database), :name view-nm)]
               ;; and its columns' :base_type should have been identified correctly
-            (is (= [{:name "numeric_col",   :database_type "numeric(10,2)",         :base_type :type/Decimal}
-                    {:name "weird_varchar", :database_type "character varying(50)", :base_type :type/Text}]
+            (is (= [{:name "numeric_col",   :database_type "numeric",         :base_type :type/Decimal}
```
Comment (qnkhuat) [inline modules/drivers/redshift/test/metabase/driver/redshift_test.clj:244]: any idea why we lose the precision here?

File: modules/drivers/redshift/src/metabase/driver/redshift.clj [hunk 10]
```diff
@@ -76,6 +77,40 @@
...
 
+(defmethod sql-jdbc.sync/describe-fields-sql :redshift
+  [driver & {:keys [schema-names table-names]}]
+  (sql/format {:select [[:c.column_name :name]
+                        [:c.data_type :database-type]
+                        [[:- :c.ordinal_position 1] :database-position]
+                        [:c.schema_name :table-schema]
+                        [:c.table_name :table-name]
+                        [nil :database-is-auto-increment] ; only needed for actions, which redshift doesn't support yet
+                        [nil :database-required]          ; only needed for actions, which redshift doesn't support yet
```
Comment (qnkhuat) [inline modules/drivers/redshift/src/metabase/driver/redshift.clj:88]: probably don't even need to return it as these fields are optional

File: src/metabase/driver/postgres.clj
Comment (calherries) [reply src/metabase/driver/postgres.clj:1]: The subset of the  `getColumns` implementation we were using converts `timestamp with time zone` types to `timestamptz`. So previously fields with a `database-type` of `timestamptz` now have `timestamp with time zone` . Fortunately we have tests that caught this.

File: src/metabase/sync/interface.clj (hunk 8 above)
Comment (calherries) [reply src/metabase/sync/interface.clj:70]: Good question. We don't use the description anywhere. We get the description from `DatabaseMetadata` instead. https://github.com/metabase/metabase/pull/38828/files#diff-8eb3ce5146f34119083048007ec1c733c4c77e14b40be0aa84e583f1cc3ab6d1R10-R16

I removed the requirement to get the description from the interface as a way to show driver authors that they can  switch from `describe-table` to `describe-fields` without losing any information, since we only use the `fields` key from `describe-table`.

File: src/metabase/util/malli/schema.clj [hunk 11]
```diff
@@ -368,3 +368,10 @@
   (mu/with-api-error-message
    [:re u/uuid-regex]
    (deferred-tru "value must be a valid UUID.")))
+
+(defn CollectionOf
+  "Helper for creating schemas to check whether something is an instance of a collection."
+  [item-schema]
+  [:fn
+   {:error/message (format "Collection of %s" item-schema)}
+   #(and (coll? %) (every? (partial mc/validate item-schema) %))])
```
Comment (calherries) [inline src/metabase/util/malli/schema.clj:377]: This is actually dead code, but I think it's generally useful so I'm inclined to keep it around.

At the moment we have no way to type something that could be a set or a vector for example.

**Review by crisptrutski — COMMENTED** [review]
Getting notified that diffs are out of date, flushing comments before I continue.

File: src/metabase/util/malli/schema.clj (hunk 11 above)
Comment (crisptrutski) [reply src/metabase/util/malli/schema.clj:377]: Surprised it's not a built-in!

File: modules/drivers/redshift/src/metabase/driver/redshift.clj [hunk 12]
```diff
@@ -87,9 +121,15 @@
 ;; custom Redshift type handling
 
 (def ^:private database-type->base-type
-  (sql-jdbc.sync/pattern-based-database-type->base-type
-   [[#"(?i)CHARACTER VARYING" :type/Text]       ; Redshift uses CHARACTER VARYING (N) as a synonym for VARCHAR(N)
-    [#"(?i)NUMERIC"           :type/Decimal]])) ; and also has a NUMERIC(P,S) type, which is the same as DECIMAL(P,S)
```
[the remark this answers is not in the document]
Comment (calherries) [reply modules/drivers/redshift/src/metabase/driver/redshift.clj:92]: Thanks, this got deleted by accident 9edacd06c23a2af91bb32cc0d9d8577825a561eb

File: docs/developers-guide/driver-changelog.md (hunk 4 above)
Comment (calherries) [reply docs/developers-guide/driver-changelog.md]: I've optimistically changed it to 49.1 in 137c1404a72ed9b0c3d1ad20e1a21778c73ab8db

File: .clj-kondo/config.edn [hunk 13]
```diff
@@ -355,7 +355,7 @@
     metabase.sync.field-values                                      field-values
     metabase.sync.interface                                         i
     metabase.sync.schedules                                         sync.schedules
-    metabase.sync.sync-metadata.fields.fetch-metadata               fetch-metadata
+    metabase.sync.sync-metadata.fields.our-metadata                 our-metadata
```
Comment (qnkhuat) [inline .clj-kondo/config.edn:358]:
```suggestion
    metabase.sync.sync-metadata.fields.our-metadata                 fields.our-metadata
```

without this prefix, it's hard to know whether it's fields metadata or table metadata or db metadata

File: src/metabase/sync/sync_metadata/fields.clj [hunk 14]
```diff
@@ -39,29 +39,62 @@
...
+   (+ (sync-instances/sync-instances! table db-metadata (our-metadata/our-metadata table))
      ;; Now that tables are synced and fields created as needed make sure field properties are in sync.
      ;; Re-fetch our metadata because there might be somethings that have changed after calling
      ;; `sync-instances`
-     (sync-metadata/update-metadata! table db-metadata (fetch-metadata/our-metadata table))))
+      (sync-metadata/update-metadata! table db-metadata (our-metadata/our-metadata table))))
+
+(mu/defn sync-fields-for-db!
+  "Sync the Fields in the Metabase application database for a specific `table`."
+  [database :- i/DatabaseInstance]
```
Comment (qnkhuat) [inline src/metabase/sync/sync_metadata/fields.clj:73]: you can make this take `tables` and create a `[schema name] -> table` mapping that you can use inside the reducer to get the table.

Always good to reduce the number of DB calls

File: test/metabase/driver/sql_jdbc/sync/describe_table_test.clj [hunk 15]
```diff
@@ -79,6 +92,16 @@
              :json-unfolding             false}}}
          (sql-jdbc.describe-table/describe-table :h2 (mt/id) {:name "VENUES"}))))
 
+(defn- describe-fields-for-table [db table]
+  (let [driver (driver.u/database->driver db)]
+    (sort-by :database-position
+             (if (driver/database-supports? driver :describe-fields db)
+               (vec (sql-jdbc.describe-table/describe-fields driver db
```
Comment (crisptrutski) [inline test/metabase/driver/sql_jdbc/sync/describe_table_test.clj:99]: nit: it seems wasteful to materialize a full vector just then build an array and sort it, for large very N we could optimize this further. Definitely not a bottleneck though.

File: src/metabase/sync/sync_metadata/fields.clj [hunk 16]
```diff
@@ -39,29 +39,62 @@
...
+   [metabase.util.malli.schema :as ms]
+   [toucan2.core :as t2]))
 
 ;;; +----------------------------------------------------------------------------------------------------------------+
 ;;; |                                            PUTTING IT ALL TOGETHER                                             |
 ;;; +----------------------------------------------------------------------------------------------------------------+
 
-(mu/defn ^:private sync-and-update! :- ms/IntGreaterThanOrEqualToZero
-  "Sync Field instances (i.e., rows in the Field table in the Metabase application DB) for a Table, and update metadata
+ (mu/defn ^:private sync-and-update! :- ms/IntGreaterThanOrEqualToZero
```
Comment (crisptrutski) [inline src/metabase/sync/sync_metadata/fields.clj:60]: Accidental indentation?

**Review by crisptrutski — APPROVED** [review]
Nicely done 👌

File: modules/drivers/redshift/test/metabase/driver/redshift_test.clj (hunk 9 above)
Comment (calherries) [reply modules/drivers/redshift/test/metabase/driver/redshift_test.clj:244]: The subset of the implementation of `getColumns` we used previouisly uses `pg_type.typbasetype`. See
https://github.com/aws/amazon-redshift-jdbc-driver/blob/master/src/main/java/com/amazon/redshift/jdbc/RedshiftDatabaseMetaData.java#L2365C24-L2365C35
But now we're using `svv_all_columns.data_type`, which stores the precision in a separate column. 
See https://docs.aws.amazon.com/redshift/latest/dg/r_SVV_ALL_COLUMNS.html

What gives me confidence that these are roughly equivalent is that the JDBC driver implementation uses either, depending on the pattern of schemas that you select from. For example, if you select a schema pattern that corresponds to a "LOCAL_SCHEMA_QUERY", it will use `pg_type.typbasetype`, but if you specify a schema pattern that corresponds to a `NO_SCHEMA_UNIVERSAL_QUERY` it can use `svv_all_columns.data_type`.

File: docs/developers-guide/driver-changelog.md [hunk 17]
```diff
@@ -4,6 +4,16 @@ title: Driver interface changelog
 
 # Driver Interface Changelog
 
+## Metabase 0.49.1
```
Comment (qnkhuat) [inline docs/developers-guide/driver-changelog.md:7]: Oh, didn't you tell me that we don't introduce driver methods in the patch releases?

File: src/metabase/sync/sync_metadata/fields.clj (hunk 14 above)
Comment (crisptrutski) [reply src/metabase/sync/sync_metadata/fields.clj:73]: I was also bothered by the N+1 table queries, but ended up not commenting as for large N we'd rather not fetch all the tables into memory.

It seems unfortunate to me that there's a deep nest of methods to require a full table model, because in practice they're mostly using the id, name, and schema mostly. I suspect we could just add one or two more table fields to the educed rows and get rid of this query, but it would be brittle to use a partially populated instance, and refactoring all those methods to depend only on their required table fields would be pretty hairy and ugly.

File: modules/drivers/redshift/src/metabase/driver/redshift.clj [hunk 18]
```diff
@@ -87,9 +121,15 @@
...
+  (some-fn (sql-jdbc.sync/pattern-based-database-type->base-type
+            [[#"(?i)CHARACTER VARYING" :type/Text]       ; Redshift uses CHARACTER VARYING (N) as a synonym for VARCHAR(N)
+             [#"(?i)NUMERIC"           :type/Decimal]])
+           {:super       :type/*    ; (requested support in metabase#36642)
```
[the remark this answers is not in the document]
Comment (calherries) [reply modules/drivers/redshift/src/metabase/driver/redshift.clj:129]: I fear that type of comment could get out of date here. If you have any questions about `type/*` it should be documented `metabase.types` namespace. I see that it isn't, but I don't think I could do it justice if I tried right now

I'm mostly listing these types to document these types are deliberately not being handled, as opposed to accidentally not being handled. Removing the log warnings is a nice side-effect though.

File: modules/drivers/redshift/test/metabase/driver/redshift_test.clj (hunk 9 above)
[the remark this answers is not in the document]
Comment (calherries) [reply modules/drivers/redshift/test/metabase/driver/redshift_test.clj:244]: As far as I'm aware, we don't use the size / precision data from these strings anywhere.

File: src/metabase/driver.clj [hunk 19]
```diff
@@ -567,7 +582,10 @@
     :index-info
 
     ;; Does the driver support a faster `sync-fks` step by fetching all FK metadata in a single collection?
-    :describe-fks})
+    :describe-fks
+
+    ;; Does the driver support a faster `sync-fields` step by fetching all FK metadata in a single collection?
```
[the remark this answers is not in the document]
Comment (calherries) [reply src/metabase/driver.clj:587]: nice idea, since the feature name is closely coupled to the multimethod I can't see this comment becoming out of date. done 0b1742f0c8289bd97850a477f5f6aa171ac1ae30

File: src/metabase/sync/fetch_metadata.clj [hunk 20]
```diff
@@ -16,11 +16,22 @@
...
 
-(mu/defn table-metadata :- i/TableMetadata
+(mu/defn fields-metadata
+  "Effectively a wrapper for [[metabase.driver/describe-fields]] that also validates the output against the schema."
+  [database :- i/DatabaseInstance & {:as args}]
+  (cond->> (driver/describe-fields (driver.u/database->driver database) database args)
+    ;; Validate the output against the schema, except in prod.
+    ;; This is a workaround for the fact that [[mu/defn]] can't check reducible collections yet
+    (not config/is-prod?)
+    (eduction (map #(mu.fn/validate-output {} i/FieldMetadataEntry %)))))
```
[the remark this answers is not in the document]
Comment (calherries) [reply src/metabase/sync/fetch_metadata.clj:24]: Yeah I like the idea of coupling this to `metabase.util.malli.fn/fn`, and using `mu.fn/instrument-ns?` convenient. done e10924a1988450bcee4c9bf5a9c140c12edf7569

File: src/metabase/sync/sync_metadata/fields.clj (hunk 14 above)
Comment (calherries) [reply src/metabase/sync/sync_metadata/fields.clj:73]: What Chris said, I'm fetching each table because I'm avoiding fetching all the tables into memory. The good thing is app DB queries should be quite cheap compared to customer DB queries, so this is still a huge win.

In the future we can avoid the N+1 by doing a refactoring like the one I did for `sync-fks`  as I did here https://github.com/metabase/metabase/pull/39679/files#diff-e09c063342055336a262f66623a6a1783a89871e4426c33c92f8f9b5f2bcaaec

But I don't think it's worth it. That refactoring took longer than I thought because of the quirks between different app DBs.

File: test/metabase/driver/sql_jdbc/sync/describe_table_test.clj (hunk 15 above)
Comment (calherries) [reply test/metabase/driver/sql_jdbc/sync/describe_table_test.clj:99]: yeah, this is just a test helper for very small N. Not going to do this

Comment (crisptrutski) [reply test/metabase/driver/sql_jdbc/sync/describe_table_test.clj:99]: Somehow I missed this was just for a single table.

File: docs/developers-guide/driver-changelog.md (hunk 17 above)
Comment (calherries) [reply docs/developers-guide/driver-changelog.md:7]: Yeah, maybe I did say that :) I can't remember how I justified it though, probably because we've never done it before. But putting it in the 0.49.0 section doesn't make much sense.

File: src/metabase/sync/sync_metadata/fields.clj (hunk 14 above)
Comment (qnkhuat) [reply src/metabase/sync/sync_metadata/fields.clj:73]: okay, that makes sense; in that case, you need to "not" fetch all tables in metabase.sync.sync-metadata.fields line 117

File: .clj-kondo/config.edn (hunk 13 above)
Comment (calherries) [reply .clj-kondo/config.edn]: agreed. we don't use this namespace much so the extra characters doesn't have much downside 
40df070bd39e3ed15ece0bd36c50fa1c80daef2a

**Comment by calherries** [comment]
@qnkhuat can you take another look? It would be nice to get this in 49.0

**Review by qnkhuat — APPROVED** [review]
Resolve [this](https://github.com/metabase/metabase/pull/38828/files#r1524547151) then it LGTM!

**Comment by calherries** [comment]
@qnkhuat done
